include/libcyphal/transport/can/msg_tx_session.hpp

```diff
@@ -0,0 +1,123 @@
+/// @copyright
+/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
+/// Copyright Amazon.com Inc. or its affiliates.
+/// SPDX-License-Identifier: MIT
+
+#ifndef LIBCYPHAL_TRANSPORT_CAN_MSG_TX_SESSION_HPP_INCLUDED
+#define LIBCYPHAL_TRANSPORT_CAN_MSG_TX_SESSION_HPP_INCLUDED
+
+#include "delegate.hpp"
+#include "libcyphal/transport/msg_sessions.hpp"
+#include "libcyphal/transport/contiguous_payload.hpp"
+
+#include <canard.h>
+
+#include <numeric>
+
+namespace libcyphal
+{
+namespace transport
+{
+namespace can
+{
+
+/// Internal implementation details of the CAN transport.
+/// Not supposed to be used directly by the users of the library.
+///
+namespace detail
+{
+class MessageTxSession final : public IMessageTxSession
+{
+    // In use to disable public construction.
+    // See https://seanmiddleditch.github.io/enabling-make-unique-with-private-constructors/
+    struct Tag
+    {
+        explicit Tag()  = default;
+        using Interface = IMessageTxSession;
+        using Concrete  = MessageTxSession;
+    };
+
+public:
+    CETL_NODISCARD static Expected<UniquePtr<IMessageTxSession>, AnyError> make(TransportDelegate&     delegate,
+                                                                                const MessageTxParams& params)
+    {
+        auto session = libcyphal::detail::makeUniquePtr<Tag>(delegate.memory(), Tag{}, delegate, params);
+        if (session == nullptr)
+        {
+            return MemoryError{};
+        }
+
+        return session;
+    }
+
+    MessageTxSession(Tag, TransportDelegate& delegate, const MessageTxParams& params)
+        : delegate_{delegate}
+        , params_{params}
+        , send_timeout_{std::chrono::seconds{1}}
+    {
+    }
+
+private:
+    // MARK: ITxSession
+
+    void setSendTimeout(const Duration timeout) final
+    {
+        send_timeout_ = timeout;
+    }
+
+    // MARK: IMessageTxSession
+
+    CETL_NODISCARD MessageTxParams getParams() const noexcept final
+    {
+        return params_;
+    }
+
+    CETL_NODISCARD cetl::optional<AnyError> send(const TransferMetadata& metadata,
+                                                 const PayloadFragments  payload_fragments) final
+    {
+        // libcanard currently does not support fragmented payloads (at `canardTxPush`).
+        // so we need to concatenate them when there are more than one non-empty fragment.
+        // See https://github.com/OpenCyphal/libcanard/issues/223
+        //
+        const transport::detail::ContiguousPayload contiguous_payload{delegate_.memory(), payload_fragments};
+        if ((contiguous_payload.data() == nullptr) && (contiguous_payload.size() > 0))
+        {
+            return MemoryError{};
+        }
+
+        const TimePoint deadline = metadata.timestamp + send_timeout_;
+        const auto deadline_us   = std::chrono::duration_cast<std::chrono::microseconds>(deadline.time_since_epoch());
```

Comment:
I've asked for clarification via chate that the deadline object was, ultimately, received from the media layer otherwise time-since-epoch will require a call to a system clock that might not exist. Here's my query:

> Are we always getting concrete chrono clocks from media layer abstractions? I'm concerned that we may be relying on a globally-available, concrete clock implemented in the standard library. This isn't something we can expect all embedded systems to support.

Reply:
>Are we always getting concrete chrono clocks from media layer abstractions? I'm concerned that we may be relying on a globally-available, concrete clock implemented in the standard library. This isn't something we can expect all embedded systems to support.

Please refer here, section "time primitives": https://github.com/OpenCyphal-Garage/libcyphal/blob/main/docs/design/readme.md#time-primitives

Reply:
Also, `time_since_epoch` is a simple getter that does not reach out to the clock itself.